### ISSUE TITLE
"Ask for Insights on first run." breaks automation.

### ISSUE THREAD

**Issue by tthew** [bower/bower#1102]
When issuing `bower install` for the first time even with the `-s` option present bower asks  'May bower anonymously report usage statistics to improve the tool over time? (Y/n)' which breaks automated tools not expecting interaction (i.e. Travis).

Would be great if we could provide an answer as a part of the command or at least for the `-s` option to be respected.

**Comment by thanpolas** [comment]
Just wanted to express my frustration with this.

Which one is it now if we want an automated build flow like the one we used to have since ever?

```
1. bower --config.analytics=false install
2. bower --config.interactive=false install
3. bower --config.interactive=false --config.analytics=false install
```

??????

**Comment by benschwarz** [comment]
Setting an environment variable of `CI=true` on your CI server is a good start. Bower respects that, as do other tools.

**Comment by thanpolas** [comment]
@benschwarz CI is not the only environment where builds are performed.

Still waiting on an answer on how to properly, unobtrusively, silently invoke `bower install`. Setting an ENV var is a different approach and frankly not fit for our use case.

**Comment by benschwarz** [comment]
@thanpolas env vars can be added inline to one-off processes too. Not sure how that won't work for you, but happy to step through it. 

Otherwise, I recommend reading the source to see. Doesn't seem like such a mystery to me.

**Comment by thanpolas** [comment]
> Otherwise, I recommend reading the source to see. Doesn't seem like such a mystery to me.

Sure, why bother with documentation or a plain answer. On it, starting to read the codebase...

**Comment by sheerun** [comment]
@thanpolas 

- `CI=true` is equivalent to passing `--config.interactive=false`
- attaching a pipe to bower is equivalent to `--config.interactive=false`
- `--config.interactive=false` causes `--config.analytics=false` to be a default

The answer is, use one of the following:

```bash
CI=true bower install
bower install --config.interactive=false
```

**Comment by sheerun** [comment]
Documentation [was already pretty clear on it](https://github.com/bower/bower.github.io/blob/6affec4d617b2a115eb205056490a964786fb6c1/docs/api.md#running-on-a-continuous-integration-server).

I've additionally documented exact behavior of non-interactive mode: https://github.com/bower/bower.github.io/commit/19b99340e0cc6f270dcd600d196a230d08144d61

@desandro Could you review it linguistically?

**Comment by sheerun** [comment]
Small mistake:
https://github.com/bower/bower.github.io/commit/f602fbd4cb5f7e802384c05f924a236faceddb13

**Comment by desandro** [comment]
lgtm

**Comment by tonypujals** [comment]
@sheerun nice! :+1:

**Comment by waynebloss** [comment]
This is still messing me up today. Why did you have to put marketing in an automation tool?

**Comment by zjacreman** [comment]
Just ran into this issue while writing a Vagrant provisioning script. This default behavior is *intensely* annoying. It's even more annoying that I had to Google how to resolve it. Why isn't this flag documented in `bower --help`?

**Comment by benschwarz** [comment]
> This is still messing me up today. Can't you get usage statistics from your server?

No, we're not able to get everything that we want. What can we do to make things better for you?

> Why isn't this flag documented in bower --help?

I'm not sure! Would it be too much to list every single configuration option in `--help`? 

@waynebloss @zjacreman

**Comment by zjacreman** [comment]
I'd say any config flags that make it into the [API docs](bower.io/docs/api/) should be documented at the primary point of contact with the API.

The comparison was made earlier to `apt-get -y`.  You don't need any other resource but `--help` to figure out how to turn off apt prompting.

**Comment by waynebloss** [comment]
@benschwarz I'd rather have you *not ask* to collect analytics at all, like most other command line tools that I use. NPM seems to work just fine without asking me for them.

Here's the list of [collected data](https://github.com/yeoman/insight#collected-data) from [your docs](http://bower.io/docs/config/#analytics). Could you tell me what items you cannot get on the server and why you need them?

- The version of the module that implements Insight
- Module commands (e.g. install / search)
- Name and version of packages involved with command used
- Version of node.js & OS for developer debugging
- A random & absolutely anonymous ID

Also, do you collect and store my IP Address on the server side?

Thanks.

**Comment by waynebloss** [comment]
@benschwarz  Please also consider these points:

(*TL;DR: Things are not OK with this feature. People don't like it *and* you broke things to have it, which makes it seem like there is a hidden agenda here.*)

- Searching for the terms: [bower statistics](https://www.google.com/search?q=bower+statistics), you see mostly bugs and complaints.
- Nobody seems to be giving you analytics anyway, going by the low numbers seen at [bower.io/stats](http://bower.io/stats/).
- Given that nobody seems to be using this bug-causing, breaking-change of a feature - it makes me wonder why the bower team is fighting so hard against the seeming majority of people who don't want it.

This all seems very untrustworthy to me and I'll probably jump ship as it is practical. I might just be one guy but I'm representative of many and people in general (and developers in particular) are pretty sick of having their tech constantly trying to track and analyze everything we do under the guise of "making the product better".

**Comment by waynebloss** [comment]
@benschwarz I guess my final point would be this: Your documentation says it's *opt-in* but it's really not because we have to do all this work to turn off the feature. If it were *actually* opt-in, you would just display a message about it and tell people to run a different command to enable it or tell them how to turn it on.

The extra work that must be done whenever the analytics question feature is activated at an unexpected time is this:

1. Figure out what is causing the problem.
2. Go and make changes to code and/or configuration files.

That does not sound like an opt-in. That is a hard opt-out. You're not really introducing an opt-in feature when at the same time you introduce a new opt-out anti-feature.

**Comment by benschwarz** [comment]
tldr; Analytics suck support-wise, and frustrates everyone ever involved in this conversation. I want analytics to be printed out and shot at the sun. 🚀☀️💥

---

@waynebloss, let me take a second to remind you that your insinuation that the bower team has done something unjust here is slightly upsetting. 
We're predominantly volunteers, many of us are independents; Let me reassure you that there is no ill-will, nor hidden agenda. We get plenty of comments from people like you, with the same tone, and to be honest, it *is* wearing. 

Analytics _were_ added to improve Bower, and the ecosystem around it – to help the team understand what people were doing with it. Given its rate of growth from nothing to now, theres no wonder we wanted to know what was happening out there. Specific items aside — just how many people were using it? How often? Daily? Multi-daily? 

Without something like insights, we'd never have been able to open the data up to new contributors like @rayshan, who built [Bower stats](http://bower.io/stats/) as a side project. 

That said, you've made one interesting point: 

> Given that nobody seems to be using this bug-causing, breaking-change of a feature - it makes me wonder why the bower team is fighting so hard against the seeming majority of people who don't want it.

_No one_ is fighting *for* the feature — Since it was added people have had all the issues that you've mentioned. It sucks. It sucks support-wise and it sucks for people who _choose_ to deploy or server-side-build with Bower.

When installed, Bower does ask if you'd like to submit analytics, and has numerous, documented methods for disabling it if you made the wrong choice or didn't answer. We made Bower ask, because it was the right thing to do. Turns out, it is incredibly annoying. 

I know too well that a silently hanging script on the server is a total pain in the ass to find and fix, and it wastes time too. I totally understand your frustration. 

I think it'd be a lot easier to pull out analytics entirely (or, find someone to donate their time to replacing it with something else). If you want to make that very pull request, I'd be happy to review and campaign for it to get through.  

@bower/bower, thoughts?

**Comment by desandro** [comment]
If we, the Bower team, are not actively using analytics to drive decisions, then let's remove it.

**Comment by jokeyrhyme** [comment]
@benschwarz I understand that collecting anonymous usage data can help with development decisions. I think just making this a non-interactive message that explains how to opt-in without waiting for a "y" or "n" would be an immense improvement. It might also mean you can display the message more often, and in more circumstances, to get more eyeballs on it.

**Comment by waynebloss** [comment]
@benschwarz, @desandro and @bower/bower team - Sorry for the insinuation, I was just giving you my point of view as an outsider. I don't know anything about the people who built this or their motivations. Also, maybe I'm paranoid due to the fact that this battle against everything tracking me is one that I have to fight every day. Anecdotally, I think a lot of people think like me but I could be wrong (*because I don't have those stats!* :smile: )

If it's useful to anyone, why not just take out the question and let it be enabled via config? You could just print out a message that says something like "Please enable analytics to help our project by doing x. Read about how we're not invading your privacy here..."

**Comment by waynebloss** [comment]
IMO, if you really wanted those stats, I think you'd get more people with a short, polite but persistent nag message.

**Comment by thanpolas** [comment]
> @benschwarz I'd rather have you not ask to collect analytics at all, like most other command line tools that I use. NPM seems to work just fine without asking me for them.

Nooooo!! Bower is special! Bower is super special, it is not anything like the other scripts! The whole lot of the other open source packages, libraries, scripts and whatnot are really peasants compared to Bower, least would they dare to ask users for collecting their data and hanging deployment scripts around the globe losing us collectively thousands of work hours and millions of USD on debugging.

I believe we should remove the flags to turn off interactive all together and go all in, make the web and devops industry crawl to a stop, because, bower.

And then go on and force ask questions about the developer's health, well being and if they woke up well today, then we can also provide life coaching services apart from package management which is what we always really wanted to do.

**Comment by brainfish** [comment]
This whole conversation smacks of Microsoftian-level user hostile design from the Bower team.  Your community has been unequivocally clear for months that they _hate_ this, yet you refuse to address it despite great alternative suggestions.  I've already done my time digging through my CI job late in the evening finding out why the F bower decided to break it all of the sudden, and golly I'd like that time back.

Regardless of the outcome of this particular debate I feel you have done irreparable harm to your credibility from the perspective of one of your users, and my team and others we work with have begun the P.I.T.A. work of moving our package management to npm.

**Comment by benschwarz** [comment]
@brainfish, Do you feel good for posting that? I hope you do. Did you or your team ever consider contributing rather than deciding to post something like that? No? … I didn't think so. 

I plan to make a strike on changing this, because I know it frustrates people. It doesn't _personally_ bother me (because I just use the flags, like the docs explain), but I want to do it for people. 

Like I've said before, if someone wants to make the first strike on this, I'll be the first person ready to support you. Comments like @brainfish's don't help anyone. Open source people, open source.

**Comment by mex** [comment]
@benschwarz, take a look at #1921 (primarily the concept of it). That would at least make the tool less likely to break automation and cause less friction.

**Comment by arcreative** [comment]
@mex First of all.... what's up, buddy?!  

I'm not sure I like the idea that it perpetually adds 10 seconds to all affected workflows, which gives the impression that a) the process is just slow, and b) nothing really needs to be addressed.  This is like failing to check under the hood of your car when it feels less powerful, instead of being forced to check when the engine blows up.

@benschwarz That said, I'm not sure I agree with your expectation the community that we provide a PR that fixes something that was broken intentionally.  The community has spoken quite loudly, and I think we would all prefer that analytics die completely if it they can't get out of our way.  Collecting stats CANNOT prevent the tool from working properly, which it's currently doing.  

Part of the problem is that this was a HUGE breaking change.  If this were a brand new tool that had analytics included from the beginning and everyone hopped in without reading the readme, the burden is on them to fix it.  But when you enforce a certain standard of reliability and introduce something crazy like this, people get upset, and I can't say I blame them.  

So again, can we make this an opt-in process?  I don't think anyone had an issue with that.  If you have an issue with it, can we hear what that is?

**Comment by arcreative** [comment]
Or what if analytics were enabled, the option to disable is retained, but the interactive prompt was eliminated?  I wouldn't mind if it happily reported statistics without asking me, as long as it doesn't blow up my automation.

**Comment by benschwarz** [comment]
> So again, can we make this an opt-in process? I don't think anyone had an issue with that. If you have an issue with it, can we hear what that is?

I've said it before — I don't have *any* issue in changing the behaviour. I donate my time to projects like bower when I can. Its not my responsibility to do anything here.  I was going to make some changes last week, until a comment on this thread made me want to throw my computer into space :night_with_stars: 

Lets get a change through and forget this ever happened :muscle:

**Comment by thanpolas** [comment]
> Like I've said before, if someone wants to make the first strike on this, I'll be the first person ready to support you. Comments like @brainfish's don't help anyone. Open source people, open source.

@benschwarz thank you on the lesson. This is open source, discussing, no, rather, asking, for the permission of the maintainers of the repository to remove the feature. That's how it's done.

1. You announce/ask for something you want / intend to do
1. The maintainers refuse the change
1. You ask again
1. The maintainers refuse the change
1. Bunch of people go wild with the issue
1. You play the open source card
1. Somebody get's to do the 10' minute patch

**Comment by benschwarz** [comment]
I'm unsubscribing from this issue.

**Comment by mex** [comment]
@arcreative, you know, work all the time as always. I agree with you though that this isn't the most optimal solution, but it is at least better than the alternative. I did think about simply suggesting a `git revert`, which would suit me just fine as well.

I think skipping the prompt and defaulting to opt in would be a big mistake, because a lot of people care deeply about what data are shared from their systems.

**Comment by arcreative** [comment]
@thanpolas Thanks for being so polite/constructive and chasing away the only maintainer that was willing to help us with this.  I'm glad you seem to be so clear on the open source process--didn't realize it entailed being an arrogant ass about everything, I'll be sure to employ this tactic in the future, since it seems to work _so well_.

His suggestion was to SUBMIT A PR for it, not insult him until he created a PR for you.

**Comment by mex** [comment]
Maybe we can end this discussion with this: https://github.com/bower/bower/pull/1923

This both adds a check for `process.env.CI` and adds a fallback on 30 sec, if no response is given to the prompt.

**Comment by thanpolas** [comment]
@arcreative i'm afraid you got it all backwards. Bower is not reliant on a single entity, here's where open source comes in. If the issue is the PR i'd gladly submit it myself.

The problem here is that Team Bower would not accept such a PR. 

If that's not the case, then mea culpa and i'll promptly submit a PR ripping off the whole analytics system from its roots.

**Comment by waynebloss** [comment]
I like the idea of just removing analytics altogether because nobody here voiced any real need for them to be collected in the first place.

The only purpose that I could gather is that they're needed for some website that we're not sure if anyone is using.

**Comment by waynebloss** [comment]
Adding a timeout to "fix" this problem seems like a terrible idea. Could someone please explain to me why we need these analytics so badly?

Why not just make this thing truly opt-in by removing the prompt altogether?

**Comment by sheerun** [comment]
I don't like the way @thanpolas behaves in this thread, but here is my proposal anyway:

Personally, I agree that:

- Most of the command statistics (except maybe install counts) are not useful
- "Top packages" is of questionable use too, because installs are counted only install directly, like `bower install jquery`. They don't show total download/star count for modules. Ideally they should show trending repositories within packages registered in Bower registry (something like https://github.com/explore)
- "Popularity around the world" section seems not useful as well

I think we need http://bower.io/stats/ to show that bower is still alive (for example to show what versions of bower are still installed and used), and to show pulse of packages registered in bower registry (for example show "Top 50 starred repositories in last 7 days)").

Because most of packages in Bower's public registry come from GitHub, I think we could remove analytics from Bower if we could retrieve following counts:

- npm installs of bower, per version
- daily number of new stars for each "component" registered in bower

And re-implement http://bower.io/stats using these statistics

I hope core team agrees with it and I hope someone will donate his/her time to make it happen :)

**Comment by sheerun** [comment]
If course if the things I mentioned are done, we'd remove analytics from Bower, completely.

**Comment by samccone** [comment]
Can everyone on the thread make sure to treat other people working on this with respect, I am pretty ashamed at the level of discourse on this issue.  

I know it is easy to talk down and shame people, but honest mistakes are made in software. 

Unless you are going to help please try and be positive, and thank @sheerun for donating his time to fix this.

Thanks you @sheerun !!!

**Comment by sheerun** [comment]
We need to wait for @npm to publish per-version download counts (it's extremely useful statistic for any package owner, it's shame it's not available), but otherwise http://bower.io/stats is updated and #1102 ready to merge.
